docs/topics/jvm-api-guidelines-backward-compatibility.md

```diff
@@ -267,7 +267,7 @@ Let's change the return type of the "library" function `x()` from `Number` to `I
 fun x(): Int = 3
 ```
 
-And recompile only the client: `kotlinc client.kt`. `ClientKt` doesn't work as expected anymore. It doesn't print `3` 
+And recompile only the library: `kotlinc library.kt`. `ClientKt` doesn't work as expected anymore. It doesn't print `3` 
```

Comment:
I don't think it's a typo or a mistake. It demonstrates that if you change the return type of a function in your **library**, the **clients** will need to re-compile **their** code against the new version of your **library** code.

The client code was compiled to call the function that returns `Number`. However, because the return type was changed from `Number` to `Int`, such function doesn't exist anymore, so it fails with `NoSuchMethodError`. This pretty much happens because java is dynamically linked.

I'll leave it to @asm0dey, maybe we could add some details to this section so that it doesn't raise such questions

Reply:
Yes, @IgnatBeresnev, it's correct. The backward compatibility issues mainly arise during split compilation, which we emulate in this example

Reply:
I think it depends on what do you compile the client.kt against, the source code (which was updated) or the binary (which is outdated). If compiled against the source code the error message should be `java.lang.Integer` instead of `java.lang.Number`, because the client was compiled against the version with the `Int` as return type. If compiled against the binary, there should no be any error, because the binary did not change and the client code was also not changed.